research/conditional/utils/model_utils.py

```diff
@@ -456,11 +458,31 @@ def get_ff_layer(args):
             llm.FeedForward(*parallel_ff_args),
         )
     elif args.ff_mode == "token_choice":
+        if args.token_choice_logic == "relu":
+            expert_logic = lambda: ExpertRelu(
+                dmodel=args.dmodel,
+                n_experts=args.n_experts,
+                expert_size=args.expert_size,
+                init_scale=args.init_scale,
+                init_type=args.init_type,
+            )
+        elif args.token_choice_logic == "swi_glu":
+            expert_logic = lambda: ExpertSwiGLU(
+                dmodel=args.dmodel,
+                n_experts=args.n_experts,
+                expert_size=args.expert_size,
+                init_scale=args.init_scale,
+                init_type=args.init_type,
+            )
+        else:
+            raise NotImplementedError(
+                f"Token choice logic {args.token_choice_logic} not implemented"
+            )
         return_fn = lambda: TokenChoiceFF(
             dmodel=args.dmodel,
             n_experts=args.n_experts,
-            expert_size=args.expert_size,
             capacity_factor=args.capacity_factor,
+            expert_logic=expert_logic(),
```

Comment:
All in all I am not sure if some other name would not be better? Maybe `expert_inner_function` or something along the lines. Not adamant on that one, but might make the code easier to understand for a newcomer.